File: 868.Binary-Gap/binaryGap.cpp
/*
 * 位运算
 */
class Solution {
public:
    int binaryGap(int N) {
        int maxd = 0, d = 0;
        bool count = false;//计数
        
        while(N){
            if(N&1){//判断末尾是0还是1：如果是1
                count = true;
                maxd = max(maxd, d);
                d = 0;//重置d
            }
            if(count)//1和1之间有间隔
                d++;
            N = N>>1;
        }
        
        return maxd;
    }
};

File: 868.Binary-Gap/binaryGap.py
"""
要尽可能多用内置函数；
"""

class Solution:
    def binaryGap(self, N):
        """
        :type N: int
        :rtype: int
        """
        binN = bin(N)[2:]
        if binN.count('1') <= 1:
            return 0
        start = binN.find('1')
        res = 0
        for i in range(start+1, len(binN), 1):
            if binN[i] == '1':
                res = max(res, i-start)
                start = i#start要及时更新；只要找到1，就更新
        
        return res
